chap_core/geoutils.py

```diff
@@ -0,0 +1,112 @@
+'''
+Utility functions for working with geometries.
+'''
+import io
+
+from .geometry import Polygons
+from .api_types import FeatureModel
+
+# def count_coordinates(feature : FeatureModel):
+#     # TODO: not sure if needed... 
+#     geom = feature.geometry
+
+#     geotype = geom.type
+#     coords = geom.coordinates
+
+#     if geotype == "Point":
+#         xs = [coords[0]]
+#     elif geotype in ("MultiPoint","LineString"):
+#         xs = (x for x,y in coords)
+#     elif geotype == "MultiLineString":
+#         xs = (x for line in coords for x,y in line)
+#     elif geotype == "Polygon":
+#         xs = (x for ring in coords for x,y in ring)
+#     elif geotype == "MultiPolygon":
+#         xs = (x for poly in coords for ring in poly for x,y in ring)
+    
+#     return len(list(xs))
+
+def feature_bbox(feature : FeatureModel):
+    geom = feature.geometry
+
+    geotype = geom.type
+    coords = geom.coordinates
+
+    if geotype == "Point":
+        x,y = coords
+        bbox = [x,y,x,y]
+    elif geotype in ("MultiPoint","LineString"):
+        xs, ys = zip(*coords)
+        bbox = [min(xs),min(ys),max(xs),max(ys)]
+    elif geotype == "MultiLineString":
+        xs = [x for line in coords for x,y in line]
+        ys = [y for line in coords for x,y in line]
+        bbox = [min(xs),min(ys),max(xs),max(ys)]
+    elif geotype == "Polygon":
+        exterior = coords[0]
+        xs, ys = zip(*exterior)
+        bbox = [min(xs),min(ys),max(xs),max(ys)]
+    elif geotype == "MultiPolygon":
+        xs = [x for poly in coords for x,y in poly[0]]
+        ys = [y for poly in coords for x,y in poly[0]]
+        bbox = [min(xs),min(ys),max(xs),max(ys)]
+    return bbox
+
+def render(polygons : Polygons):
+    import geopandas as gpd
```

Comment:
These imports are only needed for this functions so keeping them inside. But not sure if there are other conventions you use elsewhere in chap-core.